In [1]:
import pandas as pd
import os
os.chdir("..")

In [3]:
body_train = pd.read_csv('fnc-1/train_bodies.csv')
body_train.head()

Body ID                                        articleBody
0        0  A small meteorite crashed into a wooded area i...
1        4  Last week we hinted at what was to come as Ebo...
2        5  (NEWSER) – Wonder how long a Quarter Pounder w...
3        6  Posting photos of a gun-toting child online, I...
4        7  At least 25 suspected Boko Haram insurgents we...

In [4]:
stance_train = pd.read_csv('fnc-1/train_stances.csv')
stance_train.head()

Headline  Body ID     Stance
0  Police find mass graves with at least '15 bodi...      712  unrelated
1  Hundreds of Palestinians flee floods in Gaza a...      158      agree
2  Christian Bale passes on role of Steve Jobs, a...      137  unrelated
3  HBO and Apple in Talks for $15/Month Apple TV ...     1034  unrelated
4  Spider burrowed through tourist's stomach and ...     1923   disagree

In [18]:
passage = body_train.articleBody[0]

In [27]:
import os
os.environ['NLTK_DATA'] = r'D:\NLP_PJ\nltk_data'

In [6]:
import nltk
from nltk.tokenize import sent_tokenize
# nltk.download('punkt', download_dir=r'D:\nltk_data')

In [30]:
sentences = sent_tokenize(passage)

In [31]:
sentences

["A small meteorite crashed into a wooded area in Nicaragua's capital of Managua overnight, the government said Sunday.",
 "Residents reported hearing a mysterious boom that left a 16-foot deep crater near the city's airport, the Associated Press reports.",
 'Government spokeswoman Rosario Murillo said a committee formed by the government to study the event determined it was a "relatively small" meteorite that "appears to have come off an asteroid that was passing close to Earth."',
 'House-sized asteroid 2014 RC, which measured 60 feet in diameter, skimmed the Earth this weekend, ABC News reports.',
 'Murillo said Nicaragua will ask international experts to help local scientists in understanding what happened.',
 'The crater left by the meteorite had a radius of 39 feet and a depth of 16 feet,  said Humberto Saballos, a volcanologist with the Nicaraguan Institute of Territorial Studies who was on the committee.',
 'He said it is still not clear if the meteorite disintegrated or was bu

In [32]:
import pickle as pkl

In [33]:
body_train['sentences'] = body_train['articleBody'].apply(sent_tokenize)

In [7]:
stance_train['sentences'] = stance_train['Headline'].apply(sent_tokenize)

In [34]:
body_train

Body ID                                        articleBody  \
0           0  A small meteorite crashed into a wooded area i...   
1           4  Last week we hinted at what was to come as Ebo...   
2           5  (NEWSER) – Wonder how long a Quarter Pounder w...   
3           6  Posting photos of a gun-toting child online, I...   
4           7  At least 25 suspected Boko Haram insurgents we...   
...       ...                                                ...   
1678     2528  Intelligence agencies hunting for identity of ...   
1679     2529  While Daleks "know no fear" and "must not fear...   
1680     2530  More than 200 schoolgirls were kidnapped in Ap...   
1681     2531  A Guantanamo Bay prisoner released last year a...   
1682     2532  ANN ARBOR, Mich. – A pizza delivery man in Mic...   

                                              sentences  
0     [A small meteorite crashed into a wooded area ...  
1     [Last week we hinted at what was to come as Eb...  
2     [(NEWSER) – Wonder how long a Quarter Pounder ...  
3     [Posting photos of a gun-toting child online, ...  
4     [At least 25 suspected Boko Haram insurgents w...  
...                                                 ...  
1678  [Intelligence agencies hunting for identity of...  
1679  [While Daleks "know no fear" and "must not fea...  
1680  [More than 200 schoolgirls were kidnapped in A...  
1681  [A Guantanamo Bay prisoner released last year ...  
1682  [ANN ARBOR, Mich. – A pizza delivery man in Mi...  

[1683 rows x 3 columns]

In [35]:
pkl.dump(body_train, open('sentences/body_train.pkl','wb+'))

In [46]:
pkl.dump(stance_train, open('sentences/stance_train.pkl','wb+'))

In [38]:
body_test = pd.read_csv('fnc-1/test_bodies.csv')
body_test.head()

Body ID                                        articleBody
0        1  Al-Sisi has denied Israeli reports stating tha...
1        2  A bereaved Afghan mother took revenge on the T...
2        3  CNBC is reporting Tesla has chosen Nevada as t...
3       12  A 4-inch version of the iPhone 6 is said to be...
4       19  GR editor’s Note\r\n\r\nThere are no reports i...

In [8]:
stance_test = pd.read_csv('fnc-1/test_stances_unlabeled.csv')
stance_test.head()

Headline  Body ID
0  Ferguson riots: Pregnant woman loses eye after...     2008
1  Crazy Conservatives Are Sure a Gitmo Detainee ...     1550
2  A Russian Guy Says His Justin Bieber Ringtone ...        2
3  Zombie Cat: Buried Kitty Believed Dead, Meows ...     1793
4  Argentina's President Adopts Boy to End Werewo...       37

In [9]:
stance_test['sentences'] = stance_test['Headline'].apply(sent_tokenize)

In [12]:
stance_train.head()['sentences'][0]

["Police find mass graves with at least '15 bodies' near Mexico town where 43 students disappeared after police clash"]

In [14]:
import pickle as pkl

In [18]:
pkl.dump(stance_train, open('sentences/stance_train.pkl','wb'))

In [19]:
pkl.dump(stance_test, open('sentences/stance_test.pkl','wb'))

In [43]:
body_test['sentences'] = body_test['articleBody'].apply(sent_tokenize)

In [44]:
pkl.dump(body_test, open('sentences/body_test.pkl', 'wb+'))

In [3]:
import pandas as pd
import pickle as pkl
import torch
import numpy as np

c:\Users\Song\.conda\envs\myenv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [11]:
head_train = pkl.load(open('/gpt_embeddings/stance_train.pkl','rb'))

In [12]:
body_train = pkl.load(open('/gpt_embeddings/body_train.pkl','rb'))

In [ ]:
fnc_train = pd.merge('')

In [17]:
fnc_train = head_train.merge(body_train, on='Body ID', suffixes=['_head', '_body'])

In [23]:
fnc_train['embeddings_body'] = fnc_train['embeddings_body'].apply(torch.tensor)
fnc_train['embeddings_head'] = fnc_train['embeddings_head'].apply(torch.tensor)

In [78]:
pkl.dump(fnc_train, open('/gpt_embeddings/fnc_train.pkl','wb'))

In [27]:
head_test = pkl.load(open('/gpt_embeddings/stance_test.pkl','rb'))
body_test = pkl.load(open('/gpt_embeddings/body_test.pkl','rb'))

In [ ]:
fnc_test = head_test.merge(body_test, on='Body ID', suffixes=['_head', '_body'])

In [30]:
fnc_test['embeddings_body'] = fnc_test['embeddings_body'].apply(torch.tensor)
fnc_test['embeddings_head'] = fnc_test['embeddings_head'].apply(torch.tensor)

In [31]:
pkl.dump(fnc_test, open('/gpt_embeddings/fnc_test.pkl', 'wb'))

In [38]:
fnc_train.iloc[13]

Headline           Gold Apple Watch Edition price? Speculators sa...
Body ID                                                          712
Stance                                                     unrelated
sentences_head     [Gold Apple Watch Edition price?, Speculators ...
embeddings_head    [[tensor(-0.2377), tensor(1.0626), tensor(-0.8...
articleBody        Danny Boyle is directing the untitled film\r\n...
sentences_body     [Danny Boyle is directing the untitled film\r\...
embeddings_body    [[tensor(0.0152), tensor(0.7478), tensor(-0.19...
Name: 13, dtype: object

In [41]:
first_shape = fnc_train['embeddings_head'].shape

In [47]:
fnc_train.iloc[13]['embeddings_head'].shape

torch.Size([2, 768])

### If the head / body has k sentences, the embedding is shape k * 768 

In [52]:
fnc_train.iloc[13]['sentences_head']

['Gold Apple Watch Edition price?', 'Speculators say at least $10,000']

In [51]:
fnc_train.iloc[13]['embeddings_head']

tensor([[-0.2377,  1.0626, -0.8609,  ...,  0.0631,  0.0337, -0.4823],
        [-0.6628,  0.3827, -0.6465,  ..., -0.0790,  0.6663, -0.1627]])

In [43]:
for i, tensor in fnc_train['embeddings_head'].iteritems():
    if tensor.shape != first_shape:
        print(f"Tensor at index {i} has a different shape: {tensor.shape}")

Tensor at index 0 has a different shape: torch.Size([1, 768])
Tensor at index 1 has a different shape: torch.Size([1, 768])
Tensor at index 2 has a different shape: torch.Size([1, 768])
Tensor at index 3 has a different shape: torch.Size([1, 768])
Tensor at index 4 has a different shape: torch.Size([1, 768])
Tensor at index 5 has a different shape: torch.Size([1, 768])
Tensor at index 6 has a different shape: torch.Size([1, 768])
Tensor at index 7 has a different shape: torch.Size([1, 768])
Tensor at index 8 has a different shape: torch.Size([1, 768])
Tensor at index 9 has a different shape: torch.Size([1, 768])
Tensor at index 10 has a different shape: torch.Size([1, 768])
Tensor at index 11 has a different shape: torch.Size([1, 768])
Tensor at index 12 has a different shape: torch.Size([1, 768])
Tensor at index 13 has a different shape: torch.Size([2, 768])
Tensor at index 14 has a different shape: torch.Size([1, 768])
Tensor at index 15 has a different shape: torch.Size([1, 768])
Te

In [56]:
torch.mean(fnc_train['embeddings_head'][0], dim=0).shape

torch.Size([768])

In [57]:
trans = lambda x: torch.mean(x, dim=0)

In [59]:
fnc_train_copy = fnc_train.copy()

In [61]:
fnc_test_copy = fnc_test.copy()

In [60]:
fnc_train_copy['embeddings_head'] = fnc_train_copy['embeddings_head'].apply(trans)
fnc_train_copy['embeddings_body'] = fnc_train_copy['embeddings_body'].apply(trans)

In [62]:
fnc_test_copy['embeddings_head'] = fnc_test_copy['embeddings_head'].apply(trans)
fnc_test_copy['embeddings_body'] = fnc_test_copy['embeddings_body'].apply(trans)

In [32]:
fnc_test

Headline  Body ID  \
0      Ferguson riots: Pregnant woman loses eye after...     2008   
1      Apple Stores to install safes to secure gold A...     2008   
2      Pregnant woman loses eye after police shoot be...     2008   
3      We just found out the #Ferguson Protester who ...     2008   
4      Police Chief In Charge of Paris Attacks Commit...     2008   
...                                                  ...      ...   
25408        A Sign That Obamacare Exchanges Are Failing     2586   
25409  Republicans call Obamacare a 'failure.' These ...     2586   
25410  CBO’s Alternate Facts Show Obamacare is Unsust...     2586   
25411                               Why Obamacare failed     2586   
25412  The success of the Affordable Care Act is a hu...     2586   

                                          sentences_head  \
0      [Ferguson riots: Pregnant woman loses eye afte...   
1      [Apple Stores to install safes to secure gold ...   
2      [Pregnant woman loses eye after police shoot b...   
3      [We just found out the #Ferguson Protester who...   
4      [Police Chief In Charge of Paris Attacks Commi...   
...                                                  ...   
25408      [A Sign That Obamacare Exchanges Are Failing]   
25409  [Republicans call Obamacare a 'failure.', Thes...   
25410  [CBO’s Alternate Facts Show Obamacare is Unsus...   
25411                             [Why Obamacare failed]   
25412  [The success of the Affordable Care Act is a h...   

                                         embeddings_head  \
0      [[tensor(0.3180), tensor(0.5948), tensor(-0.22...   
1      [[tensor(-0.0870), tensor(1.4910), tensor(-0.7...   
2      [[tensor(-0.0250), tensor(0.2495), tensor(-0.1...   
3      [[tensor(0.5442), tensor(0.7988), tensor(-0.63...   
4      [[tensor(0.0414), tensor(0.2250), tensor(0.547...   
...                                                  ...   
25408  [[tensor(0.0280), tensor(0.6648), tensor(1.067...   
25409  [[tensor(-0.4024), tensor(1.0369), tensor(0.12...   
25410  [[tensor(0.1872), tensor(0.8029), tensor(0.636...   
25411  [[tensor(-0.1593), tensor(0.5980), tensor(0.90...   
25412  [[tensor(-0.2598), tensor(0.6327), tensor(0.41...   

                                             articleBody  \
0      A RESPECTED senior French police officer inves...   
1      A RESPECTED senior French police officer inves...   
2      A RESPECTED senior French police officer inves...   
3      A RESPECTED senior French police officer inves...   
4      A RESPECTED senior French police officer inves...   
...                                                  ...   
25408  Remember how much Republicans wanted to repeal...   
25409  Remember how much Republicans wanted to repeal...   
25410  Remember how much Republicans wanted to repeal...   
25411  Remember how much Republicans wanted to repeal...   
25412  Remember how much Republicans wanted to repeal...   

                                          sentences_body  \
0      [A RESPECTED senior French police officer inve...   
1      [A RESPECTED senior French police officer inve...   
2      [A RESPECTED senior French police officer inve...   
3      [A RESPECTED senior French police officer inve...   
4      [A RESPECTED senior French police officer inve...   
...                                                  ...   
25408  [Remember how much Republicans wanted to repea...   
25409  [Remember how much Republicans wanted to repea...   
25410  [Remember how much Republicans wanted to repea...   
25411  [Remember how much Republicans wanted to repea...   
25412  [Remember how much Republicans wanted to repea...   

                                         embeddings_body  
0      [[tensor(-0.1159), tensor(0.4641), tensor(-0.0...  
1      [[tensor(-0.1159), tensor(0.4641), tensor(-0.0...  
2      [[tensor(-0.1159), tensor(0.4641), tensor(-0.0...  
3      [[tensor(-0.1159), tensor(0.4641), tensor(-0.0...  
4      [[tensor(-0.1159), tensor(0.4641)

In [63]:
pkl.dump(fnc_test_copy, open('/gpt_embeddings/fnc_test_mean.pkl', 'wb'))

In [77]:
pkl.dump(fnc_train_copy, open('/gpt_embeddings/fnc_train_mean.pkl', 'wb'))

In [73]:
# Define the label mapping
label_mapping = {'unrelated': 0, 'agree': 1, 'discuss': 2, 'disagree': 3}


In [74]:
fnc_train['Stance'] = fnc_train['Stance'].map(label_mapping)

In [76]:
fnc_train_copy['Stance'] = fnc_train_copy['Stance'].map(label_mapping)

In [ ]:
fnc_

In [75]:
fnc_train

Headline  Body ID  Stance  \
0      Police find mass graves with at least '15 bodi...      712       0   
1               Seth Rogen to Play Apple’s Steve Wozniak      712       2   
2      Mexico police find mass grave near site 43 stu...      712       0   
3      Mexico Says Missing Students Not Found In Firs...      712       0   
4      New iOS 8 bug can delete all of your iCloud do...      712       0   
...                                                  ...      ...     ...   
49967  Amazon Is Opening a Brick-and-Mortar Store in ...      464       1   
49968  Elon University has not banned the term ‘fresh...      362       1   
49969  Fake BBC News website set up to carry Charlie ...      915       1   
49970  Apple was forced to nix key health features fr...      407       2   
49971  If you want an Apple Watch, you’ll probably be...     1066       2   

                                          sentences_head  \
0      [Police find mass graves with at least '15 bod...   
1             [Seth Rogen to Play Apple’s Steve Wozniak]   
2      [Mexico police find mass grave near site 43 st...   
3      [Mexico Says Missing Students Not Found In Fir...   
4      [New iOS 8 bug can delete all of your iCloud d...   
...                                                  ...   
49967  [Amazon Is Opening a Brick-and-Mortar Store in...   
49968  [Elon University has not banned the term ‘fres...   
49969  [Fake BBC News website set up to carry Charlie...   
49970  [Apple was forced to nix key health features f...   
49971  [If you want an Apple Watch, you’ll probably b...   

                                         embeddings_head  \
0      [[tensor(0.1137), tensor(0.3261), tensor(-0.36...   
1      [[tensor(-0.0878), tensor(1.1093), tensor(-0.0...   
2      [[tensor(0.2799), tensor(0.6853), tensor(-0.44...   
3      [[tensor(0.2070), tensor(0.5203), tensor(0.455...   
4      [[tensor(-0.6000), tensor(0.3704), tensor(0.28...   
...                                                  ...   
49967  [[tensor(-0.1653), tensor(0.8541), tensor(-0.2...   
49968  [[tensor(0.7944), tensor(0.0871), tensor(0.559...   
49969  [[tensor(0.7371), tensor(0.6072), tensor(0.008...   
49970  [[tensor(0.0990), tensor(1.4528), tensor(0.410...   
49971  [[tensor(-0.0209), tensor(0.6179), tensor(0.39...   

                                             articleBody  \
0      Danny Boyle is directing the untitled film\r\n...   
1      Danny Boyle is directing the untitled film\r\n...   
2      Danny Boyle is directing the untitled film\r\n...   
3      Danny Boyle is directing the untitled film\r\n...   
4      Danny Boyle is directing the untitled film\r\n...   
...                                                  ...   
49967  Amazon, the cyber store that sells everything,...   
49968  ELON, N.C. – A recent rumor claims that Elon U...   
49969  A realistic-looking fake BBC News website has ...   
49970  The health-focused smartwatch that Apple initi...   
49971  We still don’t know the exact date the Apple W...   

                                          sentences_body  \
0      [Danny Boyle is directing the untitled film\r\...   
1      [Danny Boyle is directing the untitled film\r\...   
2      [Danny Boyle is directing the untitled film\r\...   
3      [Danny Boyle is directing the untitled film\r\...   
4      [Danny Boyle is directing the untitled film\r\...   
...                                                  ...   
49967  [Amazon, the cyber store that sells everything...   
49968  [ELON, N.C. – A recent rumor claims that Elon ...   
49969  [A realistic-looking fake BBC News website has...   
49970  [The health-focused smartwatch that Apple init...   
49971  [We still don’t know the exact date the Apple ...   

                                         embeddings_body  
0      [[tensor(0.0152), tensor(0.7478), tensor(-0.19...  
1      [[tensor(0.0152), tensor(0.7478), tensor(-0.19...  
2      [[tensor(0.0152), tensor(0.7478), tensor(-0.19...  
3  

In [79]:
fnc_train_copy

Headline  Body ID  Stance  \
0      Police find mass graves with at least '15 bodi...      712       0   
1               Seth Rogen to Play Apple’s Steve Wozniak      712       2   
2      Mexico police find mass grave near site 43 stu...      712       0   
3      Mexico Says Missing Students Not Found In Firs...      712       0   
4      New iOS 8 bug can delete all of your iCloud do...      712       0   
...                                                  ...      ...     ...   
49967  Amazon Is Opening a Brick-and-Mortar Store in ...      464       1   
49968  Elon University has not banned the term ‘fresh...      362       1   
49969  Fake BBC News website set up to carry Charlie ...      915       1   
49970  Apple was forced to nix key health features fr...      407       2   
49971  If you want an Apple Watch, you’ll probably be...     1066       2   

                                          sentences_head  \
0      [Police find mass graves with at least '15 bod...   
1             [Seth Rogen to Play Apple’s Steve Wozniak]   
2      [Mexico police find mass grave near site 43 st...   
3      [Mexico Says Missing Students Not Found In Fir...   
4      [New iOS 8 bug can delete all of your iCloud d...   
...                                                  ...   
49967  [Amazon Is Opening a Brick-and-Mortar Store in...   
49968  [Elon University has not banned the term ‘fres...   
49969  [Fake BBC News website set up to carry Charlie...   
49970  [Apple was forced to nix key health features f...   
49971  [If you want an Apple Watch, you’ll probably b...   

                                         embeddings_head  \
0      [tensor(0.1137), tensor(0.3261), tensor(-0.364...   
1      [tensor(-0.0878), tensor(1.1093), tensor(-0.07...   
2      [tensor(0.2799), tensor(0.6853), tensor(-0.441...   
3      [tensor(0.2070), tensor(0.5203), tensor(0.4550...   
4      [tensor(-0.6000), tensor(0.3704), tensor(0.280...   
...                                                  ...   
49967  [tensor(-0.1653), tensor(0.8541), tensor(-0.20...   
49968  [tensor(0.7944), tensor(0.0871), tensor(0.5595...   
49969  [tensor(0.7371), tensor(0.6072), tensor(0.0083...   
49970  [tensor(0.0990), tensor(1.4528), tensor(0.4107...   
49971  [tensor(-0.0209), tensor(0.6179), tensor(0.396...   

                                             articleBody  \
0      Danny Boyle is directing the untitled film\r\n...   
1      Danny Boyle is directing the untitled film\r\n...   
2      Danny Boyle is directing the untitled film\r\n...   
3      Danny Boyle is directing the untitled film\r\n...   
4      Danny Boyle is directing the untitled film\r\n...   
...                                                  ...   
49967  Amazon, the cyber store that sells everything,...   
49968  ELON, N.C. – A recent rumor claims that Elon U...   
49969  A realistic-looking fake BBC News website has ...   
49970  The health-focused smartwatch that Apple initi...   
49971  We still don’t know the exact date the Apple W...   

                                          sentences_body  \
0      [Danny Boyle is directing the untitled film\r\...   
1      [Danny Boyle is directing the untitled film\r\...   
2      [Danny Boyle is directing the untitled film\r\...   
3      [Danny Boyle is directing the untitled film\r\...   
4      [Danny Boyle is directing the untitled film\r\...   
...                                                  ...   
49967  [Amazon, the cyber store that sells everything...   
49968  [ELON, N.C. – A recent rumor claims that Elon ...   
49969  [A realistic-looking fake BBC News website has...   
49970  [The health-focused smartwatch that Apple init...   
49971  [We still don’t know the exact date the Apple ...   

                                         embeddings_body  
0      [tensor(0.0830), tensor(0.3415), tensor(0.6932...  
1      [tensor(0.0830), tensor(0.3415), tensor(0.6932...  
2      [tensor(0.0830), tensor(0.3415), tensor(0.6932...  
3  

In [80]:
from sklearn.model_selection import train_test_split
fnc_train_copy_, fnc_val_copy= train_test_split(fnc_train_copy, test_size=0.2, random_state=42)

In [82]:
pkl.dump(fnc_train_copy_, open('/gpt_embeddings/fnc_train_mean.pkl', 'wb'))

In [83]:
pkl.dump(fnc_val_copy, open('/gpt_embeddings/fnc_val_mean.pkl', 'wb'))

In [10]:
fnc_train_copy_ = pkl.load(open('/gpt_embeddings/fnc_train_mean.pkl', 'rb'))

In [4]:
fnc_val_copy = pkl.load(open('/gpt_embeddings/fnc_val_mean.pkl', 'rb'))

In [8]:
fnc_val_copy['Stance'].unique()

array([0, 2, 3, 1], dtype=int64)

In [13]:
s = fnc_train_copy_.groupby('Stance').count()['Headline']

In [14]:
s / s.sum()

Stance
0    0.733047
1    0.072342
2    0.177602
3    0.017010
Name: Headline, dtype: float64

In [16]:
s_  =(s.sum() / s)

In [17]:
s_ / s_.sum()


Stance
0    0.017136
1    0.173643
2    0.070729
3    0.738492
Name: Headline, dtype: float64

In [9]:
fnc_val_copy.groupby('Stance').count()

Headline  Body ID  sentences_head  embeddings_head  articleBody  \
Stance                                                                    
0           7240     7240            7240             7240         7240   
1            786      786             786              786          786   
2           1809     1809            1809             1809         1809   
3            160      160             160              160          160   

        sentences_body  embeddings_body  
Stance                                   
0                 7240             7240  
1                  786              786  
2                 1809             1809  
3                  160              160